### ISSUE TITLE
Allow access to request message via $req in on_success of samples

### ISSUE THREAD

**Issue by beccasaurus** [googleapis/gapic-generator#2759]
Some APIs return Empty

This means, in the `on_success` we have no variables available to print, we can only print literal strings.

```yaml
on_success:
- print: ["Deleted Dog."]
```

But we often, in hand-written samples, print out the "name"* of what was deleted

```yaml
# method name: DeleteDog

parameters:
  attributes:
  - parameter: name
    sample_argument_name: name

on_success:
- print:
  - "Deleted dog %s"
  - $req.name
```

> \* "name" is often the full path identifier of resources in Google APIs

### Issue

So, actually. Our samples usually pass in path parts. So the above isn't realistic.

This is what we would actually do:

```yaml
# method name: DeleteDog

parameters:
  attributes:
  - parameter: name%project
    sample_argument_name: project_id
  - parameter: name%dog
    sample_argument_name: dog_id

on_success:
- print:
  - "Deleted dog %s"
  - $req.name   # <-- name is composed. the sample uses helper method to build full path.
```

@hzyi-google for your consideration (v1.2)

/cc @munkhuushmgl @googleapis/samplegen

**Comment by hzyi-google** [comment]
This **should** just work after we let gapic-generator know `$req` refers to the request object. I'll try it out.

One slight concern. The generated sample may look like:

```java
public static void sampleDeleteDog(String projectId, String dogId) {
  // client = new Client()...
  String dogName = new ProjectDogName(projectId, dogId).toString();
  
  DeleteDogRequest request = new DeleteDogRequest().setDogName(dogName).build();
  client.deleteDog(request);
  System.out.printf("Deleted dog %s\n", request.getDogName()); // while this line could be System.out.printf("Deleted dog %s\n", dogName);
}
```

As commented inline, there's already a variable set up for the name of the dog, so it seems more concise to reuse that variable. WDYT?

However I'd like to point out that `$req.name` is way much easier to implement. So let's just go down this road if you do not prefer the other approach.

**Comment by beccasaurus** [comment]
Good point!

Wouldn't it make the most sense for code statements (`on_success:` elements) to have access to all of the locals of the scope?

Where locals = [ locals from the request setup + define'd variables in response + loop variable of current scope, if any ]

Anything in on_success should have access to:

projectId
dogId
dogName
request
And any `define`'d variables

**Comment by hzyi-google** [comment]
tl;dr: I think that's the better solution too, although we may need to refactor the generator a bit to achieve this.

===== Below are just some notes to help myself remember what needs to be done to support this feature.

So far sample config does not have enough control over how a request object is set up. For example, the local variable names are automatically picked by the generator, so on_success can't actually have access to them.

Also the ways different languages set up request objects are different :( . e.g.,

C# and Go -> set up the request object in one nested statement
Java -> sets up the request object, but in multiple statements, with a lot of local variables
Ruby, Python -> do not use request object, use flattened parameters

**Comment by hzyi-google** [comment]
Moved to backlog: https://github.com/googleapis/gapic-generator/wiki/feature-backlog
